syntax = "proto3";
package mud;

message player {
	string name = 1;
	string password_hash = 2;
	int64 id = 3;
	repeated int64 id_characters = 4;
}

message player_book {
	repeated player players = 1;
}

message enum_character{
	enum character_state {
		NONE = 0;
		WALKING = 1;
		COMBAT = 2;
	}
}

message enum_atribute_name{
	enum attribute_name {
			LIFE = 0;
			STRENGTH = 1;
			AGILITY = 2;
			INTELLIGENCE = 3;
	}
}

message enum_direction{
	enum direction {
		NORTH = 0;
		SOUTH = 1;
		EAST = 2;
		WEST = 3;
	}
	direction value = 1;
}

message character {
	string name = 1;
	int64 id = 2;
	int64 tile_id = 3;
	enum_direction facing = 4;
	repeated enum_atribute_name attributes = 5;
	enum_character state = 6;
}

message enemy {
	string name = 1;
	int64 id = 2;
	int64 tile_id = 3;
	enum_direction facing = 4;
	repeated enum_atribute_name attributes = 5;
	enum_character state = 6;
}

message character_book {
	repeated character characters = 1;
}

message enemy_book {
	repeated enemy enemies = 1;
}

enum tile_type {
	EMPTY = 0;
	WALL = 1;
	TREE = 2;
	PORTAL = 3;
}

message location {
	int64 north = 1;
	int64 south = 2;
	int64 east = 3;
	int64 west = 4;
}

enum resident_type {
	NOBODY = 0;
	CHARACTER = 1;
	ENEMY = 2;
}


message tile {
	int64 id = 1;
	tile_type type = 2;
	location neighbour = 3;
	resident_type occupant_type = 4;
	int64 occupant_id = 5;
	string mood = 6;
}

message tile_book {
	repeated tile tiles = 1;
}


